### ISSUE TITLE
MCP_Geometric on extremely large array

### ISSUE THREAD

**Issue by JoyMonteiro** [scikit-image/scikit-image#1478]
Hello,

I was trying to calculate the shortest path on an array of size (18000, 12000).
The size of the file (TIFF file) on disk is around 825 MB.

I am trying this on Ubuntu 14.10, and the latest version of scikit-image available
to pip. My computer is an i7 with 12 GB of RAM.

The strange thing is that MCP_Geometric(costs, fully_connected=True) succeeds
sometimes and fails other times. It takes up most of my memory and then releases
most of it when it succeeds. Other times, it simply keeps taking up memory and
I saw my swap usage go upto 7 GB, with python taking up some 20 GB of memory
when I checked using "top".

Am I asking for too much from MCP_Geometric, or is there something I'm missing?
I would rather not move to a specialised tool like graph-tool unless I really have to.

TIA,
Joy

**Comment by ozak** [comment]
Hi @JoyMonteiro I have used MCP_Geometric on 2-4GB rasters without a problem (as long as the machine has enough memory). I also see that construction of the MCP object and computation of optimal costs increases memory usage by a large amount (sometimes consuming up to 40-50GB memory), and then releasing them once computation is finished.

**Comment by stefanv** [comment]
Thanks, Ömer, for having a look at the MCP code.  I realize it has 
to build a fairly large internal structure, but it should not take 
up much more space than the size of the image itself.  On Windows, 
large memory allocations can be problematic.

As mentioned in another thread, the option of doing a raw calculation
always exists, which can be less memory intensive but may be slower.

**Comment by ozak** [comment]
Hi Stefan,

no problem...I have been meaning to work on this for a while.

I only work on OSX or Linux and the memory allocation/consumption is always an issue. For some of the work I am doing I had to change the image size from 40000 x 23000 to 4000 x 2300, otherwise I would not be able to run things on our cluster.

**Comment by JoyMonteiro** [comment]
Thanks, Ömer. On further checking, MCP seems to do ok for my raster, as long
as I have only one graph in memory.

Stefan, which thread are you referring to?

Thanks,
Joy

**Comment by stefanv** [comment]
I can't remember the exact thread, but I think it's in the mailing 
list archive.

A 40000 x 40000 array, even if it is stored as int32s still occupies 6
GB of RAM!  So, any temporaries or copies will be disastrous.

**Comment by JoyMonteiro** [comment]
Thanks, found it!

**Comment by ozak** [comment]
@JoyMonteiro can you post the thread number.

Stefan you are right, large rasters are a mess. Sadly, sometimes you need to work with them.

**Comment by stefanv** [comment]
You may also want to look at `dask`, which allows you to do blocked
operations on large arrays.

**Comment by JoyMonteiro** [comment]
@ozak  Here it is... https://github.com/scikit-image/scikit-image/issues/1353

**Comment by JoyMonteiro** [comment]
@stefanv sorry for returning to this so late, but I realised that the low memory version you talked
about was not the one I mentioned in the previous post. Tried looking at the mailing list archives, but searching for MCP does not show up anything of relevance.

are you talking about shortest_path() (https://github.com/scikit-image/scikit-image/blob/master/skimage/graph/spath.py) ?

**Comment by stefanv** [comment]
Yes, exactly: constructing a cost matrix and then using dynamic programming to calculate the cost.

**Comment by JoyMonteiro** [comment]
Tried going through the mailing list archives regarding this function, and it is not clear (even looking at the code) how I can route from a point to another point, both specified. I think this was also part of the attraction in introducing the MCP class later on. Any tips on how to use it for such an application?

**Comment by ozak** [comment]
@JoyMonteiro have you updated scikit-image? We have changed part of the code to try to make it more memory efficient. That should help in cases like yours or mine.

**Comment by JoyMonteiro** [comment]
oh, I have not. I'm downloading 0.11.3, is that the version you are talking about?

**Comment by JoyMonteiro** [comment]
Yes, it does make a computer suffer a little less. I was looking at mcp.pyx, and I noticed that
the cost array was being cast as float64. I don't need that kind of precision, so will there be
a significant memory saving if I make that float32? or is the major memory consumption due
to the data structures being created in __init__()?

**Comment by ozak** [comment]
It is included in the current development version as far as I know. 
As far as I can understand the code most of the memory requirements are in init(), which is where hopefully now there is some efficiency gains. I had stated looking at the parallellization issue, but have not had much time to work on it. One thingyou can do is actually convert the raster to a smaller resolution and do the computations then. I did some trials using my package gisrasters and the results are not too bad, but you can gain lots of efficiency. Another solution is to split the raster and do piecewise computations.
